src/llvm-late-gc-lowering.cpp

```diff
@@ -2105,11 +2113,54 @@ bool LateLowerGCFrame::CleanupIR(Function &F, State *S) {
                     });
                 newI->takeName(CI);
 
+                // LLVM alignment/bit check is not happy about addrspacecast and refuse
+                // to remove write barrier because of it.
+                // We pretty much only load using `T_size` so try our best to strip
+                // as many cast as possible.
+#if JL_LLVM_VERSION >= 100000
+                auto tag = CI->getArgOperand(2)->stripPointerCastsAndAliases();
+#else
+                auto tag = CI->getArgOperand(2)->stripPointerCasts();
+#endif
+                if (auto C = dyn_cast<ConstantExpr>(tag)) {
+                    if (C->getOpcode() == Instruction::IntToPtr) {
+                        tag = C->getOperand(0);
+                    }
+                }
+                else if (auto LI = dyn_cast<LoadInst>(tag)) {
+                    // Make sure the load is correctly marked as aligned
+                    // since LLVM might have removed them.
+                    // We can't do this in general since the load might not be
+                    // a type in other branches.
+                    // However, it should be safe for us to do this on const globals
+                    // which should be the important cases as well.
+                    if (isLoadFromConstGV(LI) && getLoadValueAlign(LI) < 16) {
```

Comment:
This feels like it might not apply in imaging mode? I'm not really sure how they're represented by then here.

Reply:
What do you mean? Most of what I see here are still loads of global variables much like how they are emitted by codegen though some have their alignment bit stripped off by LLVM (mostly LICM). There are also some calls but those are handled by the return alignment attribute and doesn't really get stripped off that easily... Since for the most important case (allocation of object of known type) it's really just a constant I don't think there's much else LLVM want to do with these other than moving them around.

And this is really an optimization to make convey the same info to LLVM while keeping the IR as simple as possible. As long as the information is correct, any cases that aren't captured by this will just use the fallback alignment assumption below. Given how far it progresses when I add an abort in the fallback assume branch below I think most of the cases are already captured here. The main (only?) exception is some phi node, which are probably either directly or indirectly caused by union splitting and that's the main reason I'm still keeping that branch...